digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/assetUpdater/LaneUpdater.scala

```diff
@@ -0,0 +1,414 @@
+
+    // Adjust m-values and side codes for lanes marked to be adjusted
+    if (changeSet.positionAdjustments.nonEmpty) {
+      logger.info("Saving SideCode/M-Value adjustments for lane Ids:" + changeSet.positionAdjustments.map(a => "" + a.laneId).mkString(", "))
+      saveLanePositionAdjustments(changeSet.positionAdjustments)
+    }
+  }
+
+
+
+  def updateLanes(): Unit = {
+    val allRoadLinkChanges = roadLinkChangeClient.getRoadLinkChanges()
+    updateTrafficDirectionChangesLaneWorkList(allRoadLinkChanges)
+    val (workListChanges, roadLinkChanges) = allRoadLinkChanges.partition(change => isOldLinkOnLaneWorkList(change))
+    val changeSet = handleChanges(roadLinkChanges, workListChanges)
+    updateChangeSet(changeSet)
+  }
+
+  def updateTrafficDirectionChangesLaneWorkList(roadLinkChanges: Seq[RoadLinkChange]): Unit = {
+    val oldLinkIds = roadLinkChanges.flatMap(_.oldLink).map(_.linkId)
+    val lanesOnOldLinks = laneService.fetchAllLanesByLinkIds(oldLinkIds, newTransaction = false)
+    val changesFiltered = roadLinkChanges.filterNot(change => change.changeType == RoadLinkChangeType.Add ||
+      change.changeType == RoadLinkChangeType.Remove)
+    changesFiltered.foreach(change => {
+      val oldLink = change.oldLink.get
+      val trafficDirectionChanged = isRealTrafficDirectionChange(change)
+      val oldLinkHasAdditionalLanes = lanesOnOldLinks.exists(lane => lane.linkId == oldLink.linkId && !LaneNumber.isMainLane(lane.laneCode))
+      // If traffic direction has changed and old link has additional lanes, raise road link to lane work list
+      if (trafficDirectionChanged && oldLinkHasAdditionalLanes) {
+        change.newLinks.foreach(newLink => {
+          val oldTD = oldLink.trafficDirection
+          val newTD = newLink.trafficDirection
+          val workListItem = LaneWorkListItem(0, oldLink.linkId, "traffic_direction", oldTD.value, newTD.value,
+            DateTime.now(), AutoGeneratedUsername.generatedInUpdate)
+          laneWorkListService.workListDao.insertItem(workListItem)
+        })
+      }
+    })
+  }
+
+  // If lanes from old roadLink are currently on lane work list
+  // then only process the main lanes
+  def isOldLinkOnLaneWorkList(change: RoadLinkChange): Boolean = {
+      val linkIdsOnWorkList = laneWorkListService.getLaneWorkList.map(_.linkId)
+      change.oldLink match {
+        case Some(oldLink) =>
+          linkIdsOnWorkList.contains(oldLink.linkId)
+        case None => false
+      }
+  }
+
+  def isRealTrafficDirectionChange(change: RoadLinkChange): Boolean = {
+    change.newLinks.exists(newLink => {
+      val oldOriginalTrafficDirection = change.oldLink.get.trafficDirection
+      val newOriginalTrafficDirection = newLink.trafficDirection
+      val replaceInfo = change.replaceInfo.find(_.newLinkId == newLink.linkId).get
+      val isDigitizationChange = replaceInfo.digitizationChange
+      val overWrittenTdValueOnNewLink = TrafficDirectionDao.getExistingValue(newLink.linkId)
```

Comment:
mites uudella linkillä voi olla ylikirjoitus ? Tää taitaa nyt olettaa että roadlinkproperties samuutus tehty ekana?
Kai tässä pitäisi ehkä kokeilla uudella ja vanhalla id

Reply:
olikos meillä joku tiketti taatakseen tämä?

Reply:
Ei taida erikseen tikettiä olla, pitää vään katsoa, kun näitä laitetaan ajoon, että menee oikeassa järjestyksessä. Kirjaan tämän samuutksen dokumentaatioon kunhan siihen asti kerkeän. Lambdaan voisi sitten myös jättää kommentin aiheesta kun nämä sinne lisätään.

Reply:
ei onnistu. Lambda vetää ne kaikki samaan aikaan. Teetkö tiketin asiasta. Vaatii vähän mietintää miten tää tehdään.
Varmaan step function pitää käyttää.

Reply:
Hmm joo, teen tiketin niin ei unohdu